mlflow/tracking/_tracking_service/client.py

```diff
@@ -219,30 +223,70 @@ def get_trace(self, request_id) -> Trace:
         trace_data = self._download_trace_data(trace_info)
         return Trace(trace_info, trace_data)
 
-    def search_traces(
+    def _search_traces(
         self,
         experiment_ids: List[str],
         filter_string: Optional[str] = None,
         max_results: int = SEARCH_TRACES_DEFAULT_MAX_RESULTS,
         order_by: Optional[List[str]] = None,
         page_token: Optional[str] = None,
-    ) -> PagedList[Trace]:
-        trace_infos, token = self.store.search_traces(
+    ):
+        return self.store.search_traces(
             experiment_ids=experiment_ids,
             filter_string=filter_string,
             max_results=max_results,
             order_by=order_by,
             page_token=page_token,
         )
+
+    def search_traces(
+        self,
+        experiment_ids: List[str],
+        filter_string: Optional[str] = None,
+        max_results: int = SEARCH_TRACES_DEFAULT_MAX_RESULTS,
+        order_by: Optional[List[str]] = None,
+        page_token: Optional[str] = None,
+    ) -> PagedList[Trace]:
+        def download_trace_data(trace_info: TraceInfo) -> Optional[Trace]:
+            """
+            Downloads the trace data for the given trace_info and returns a Trace object.
+            If the download fails (e.g., the trace data is missing or corrupted), returns None.
+            """
+            try:
+                trace_data = self._download_trace_data(trace_info)
+            except (MlflowTraceDataNotFound, MlflowTraceDataCorrupted) as e:
+                _logger.debug(
+                    f"Failed to download trace data for trace with request_id={e.request_id}",
+                    trace_info.request_id,
+                    exc_info=True,
+                )
+                return None
+            else:
+                return Trace(
+                    trace_info=trace_info,
+                    trace_data=trace_data,
+                )
+
+        traces = []
+        next_max_results = max_results
+        next_token = page_token
         with ThreadPoolExecutor() as executor:
-            traces = executor.map(
-                lambda ti: Trace(
-                    trace_info=ti,
-                    trace_data=self._download_trace_data(ti),
-                ),
-                trace_infos,
-            )
-        return PagedList(traces, token)
+            while len(traces) < max_results:
+                trace_infos, next_token = self._search_traces(
+                    experiment_ids=experiment_ids,
+                    filter_string=filter_string,
+                    max_results=next_max_results,
+                    order_by=order_by,
+                    page_token=next_token,
+                )
+                traces.extend(t for t in executor.map(download_trace_data, trace_infos) if t)
+
+                if not next_token:
+                    break
+
+                next_max_results = max_results - len(traces)
```

Comment:
Can we add unit test coverage for this pagination logic to make sure that this terminates?

Reply:
@dbczumar Got it, added a few more test cases. Could you take a look!